sql/index/pilosa/mapping.go

```diff
@@ -19,130 +20,181 @@ const (
 // - index name: columndID uint64 -> location []byte
 // - frame name: value []byte (gob encoding) -> rowID uint64
 type mapping struct {
-	db *bolt.DB
+	dir string
+
+	mut sync.RWMutex
+	db  *bolt.DB
+
+	clientMut sync.Mutex
+	clients   int
 }
 
-func openMapping(dir string) (*mapping, error) {
-	db, err := bolt.Open(filepath.Join(dir, mappingFileName), 0640, nil)
-	if err != nil {
-		return nil, err
-	}
+func newMapping(dir string) *mapping {
+	return &mapping{dir: dir}
+}
 
-	return &mapping{db: db}, nil
+func (m *mapping) open() {
+	m.clientMut.Lock()
+	defer m.clientMut.Unlock()
+	m.clients++
 }
 
 func (m *mapping) close() error {
+	m.clientMut.Lock()
+	defer m.clientMut.Unlock()
+
+	if m.clients > 1 {
+		m.clients--
+		return nil
+	}
+
+	m.clients = 0
+
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
 	if m.db != nil {
-		return m.db.Close()
+		if err := m.db.Close(); err != nil {
+			return err
+		}
+		m.db = nil
 	}
 
 	return nil
 }
 
-func (m *mapping) getRowID(frameName string, value interface{}) (uint64, error) {
-	var (
-		id  uint64
-		buf bytes.Buffer
-	)
-
-	enc := gob.NewEncoder(&buf)
-	err := enc.Encode(value)
-	if err != nil {
-		return id, err
+func (m *mapping) query(fn func() error) error {
+	m.mut.Lock()
+	if m.db == nil {
+		var err error
+		m.db, err = bolt.Open(filepath.Join(m.dir, mappingFileName), 0640, nil)
+		if err != nil {
+			m.mut.Unlock()
+			return err
+		}
 	}
+	m.mut.Unlock()
+
+	m.mut.RLock()
```

Comment:
ohh, here we use _Read Lock_ (I didn't catch it before. I thought it the same kind of lock, so we can reduce it). I'm wondering if it makes sense, because `putLocation` is a write/update operation.

Reply:
That mutex is only to protext `m.db`, the only place `m.db` is mutated is here so we should be safe with a read lock